### PR TITLE
Convert bgp_multipath_relax ansible playbook to pytest

### PR DESCRIPTION
**Pull request by arlakshm**
Signed-off-by: Arvindsrinivasan Lakshminarasimhan <arlakshm@microsoft.com>

### Description of PR

Summary:
Convert bgp_multipath_relax ansible playbook to pytest
Fixes # (issue)

### Type of change
New Pytest

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### How did you verify/test it?
arlakshm@3176a03c5c39:/var/repos/public/my_repo/sonic-mgmt/tests$ py.test $PYTEST_COMMON_OPTS --skip_sanity --log-file logs/test_bgp_multipath_relax.log --junitxml=results/tr.xml test_bgp_multipath_relax.py
================================================== test session starts ==================================================
platform linux2 -- Python 2.7.12, pytest-4.6.6, py-1.8.0, pluggy-0.13.0 -- /usr/bin/python
cachedir: .pytest_cache
ansible: 2.8.7
rootdir: /var/repos/public/my_repo/sonic-mgmt/tests, inifile: pytest.ini
plugins: ansible-2.2.2
collected 1 item

test_bgp_multipath_relax.py::test_bgp_multipath_relax PASSED                                                      [100%]

--------------------- generated xml file: /var/repos/public/my_repo/sonic-mgmt/tests/results/tr.xml ---------------------
=============================================== 1 passed in 15.38 seconds ===============================================
arlakshm@3176a03c5c39:/var/repos/public/my_repo/sonic-mgmt/tests
#### Any platform specific information?
None
#### Supported testbed topology if it's a new test case?
This test is supported in t1 topologies

### CODE REVIEW

**Comment by msosyak** [comment]
Please, add wrapper instead of ansible version of the test

**Comment by lguohan** [comment]
current t0 route announce does not advertise the vip prefix. I believe in order for the test to work, we need to develop a fixture that leverage the exabgp to advertise the vip prefix. I cannot find this test has that configuration.

**Comment by lguohan** [comment]
please also move the test into bgp subfolder.

**Comment by wangxin** [comment]
With PR https://github.com/Azure/sonic-mgmt/pull/1692, the vips routes can be announced to the t0 VMs that have the vips configuration.

Since the vips routes are part of the topology definition. IMO, it's OK to announce the vips routes in the fib fixture.

File: tests/test_bgp_multipath_relax.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    assert len(bgp_route_neiadv[vips_prefix]['aspath']) == 2
+
+    assert int(bgp_route_neiadv[vips_prefix]['aspath'][0]) in vips_t0 and \
+        int(bgp_route_neiadv[vips_prefix]['aspath'][1]) in vips_asn
+
+
+
+
+
+
```
Comment (wangxin) [inline tests/test_bgp_multipath_relax.py:126]: Could you please remove the extra new lines at the end of the script?

**Comment by wangxin** [comment]
Tested this PR, it passed with change #1692.

**Comment by wangxin** [comment]
retest vsimage please

File: tests/test_bgp_multipath_relax.py (hunk 1 above)
Comment (arlakshm) [reply tests/test_bgp_multipath_relax.py]: fixed in the latest commit

**Comment by arlakshm** [comment]
> please also move the test into bgp subfolder.

Moved to bgp folder

File: tests/bgp/test_bgp_multipath_relax.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    
+    logger.info("vips_prefix = {}".format(vips_prefix))
+
+    # find all paths of the prefix for test 
+    vips_t0, vips_asn = get_vips_prefix_paths(dut_t0_neigh, vips_prefix, topo_config)
+    
+    logger.info("vips_t0: {}, vips_asn: {}".format(vips_t0, vips_asn))
+
+    if len(vips_t0) <= 1:
+        pytest.fail("Did not find preconfigured multipath for the vips prefix under test")
```
Comment (tahmed-dev) [inline tests/bgp/test_bgp_multipath_relax.py:86]: can you please unify the use of assert and pytest.fail. I'd like to mention that Pytest has a method to assert called [pytest_assert](https://github.com/Azure/sonic-mgmt/blob/master/tests/common/helpers/assertions.py#L3)

**Comment by tahmed-dev** [comment]
retest vsimage please

File: tests/bgp/test_bgp_multipath_relax.py (hunk 2 above)
Comment (arlakshm) [reply tests/bgp/test_bgp_multipath_relax.py]: @tahmed-dev  Changed the assert to pytest_assert in the latest commit

File: tests/bgp/test_bgp_multipath_relax.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    pytest_assert((vips_t0 > 1), "Did not find preconfigured multipath for the vips prefix under test")
+        
+    # Get the route from the DUT for the prefix
+    bgp_route = duthost.bgp_route(prefix=vips_prefix)['ansible_facts']['bgp_route']
+    
+    logger.info("Bgp route from DUT for prefix {} is {}".format(vips_prefix,bgp_route))
+
+    # Verify found vips prefix entry in Sonic bgp routes
+    assert bgp_route[vips_prefix]['found'] ==True
```
Comment (tahmed-dev) [inline tests/bgp/test_bgp_multipath_relax.py:94]: `pytest_assert(bgp_route[vips_prefix]['found'], "BGP routes not found")`?

File: tests/bgp/test_bgp_multipath_relax.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+    # Verify total multipath match number of t0 with vips that has prefix for test
+    assert int(bgp_route[vips_prefix]['path_num']) == len(vips_t0)
```
Comment (tahmed-dev) [inline tests/bgp/test_bgp_multipath_relax.py:97]: pytest_assert?

**Comment by wangxin** [comment]
retest vsimage please

**Comment by arlakshm** [comment]
retest vsimage please

**Comment by arlakshm** [comment]
retest vsimage please

**Comment by tahmed-dev** [comment]
retest vsimage please

**Comment by arlakshm** [comment]
retest vsimage please

**Pull request merged**